### DISCUSSION TITLE
systemd-boot support and L4TLauncher

### DISCUSSION THREAD

**Discussion by igoropaniuk** [OE4T/meta-tegra#1650] [General]
I've stumbled upon multiple discussions here about integration of systemd-boot and currently trying to bring up a PoC setup, so we can replace `syslinux.cfg` and have a generic EFI boot loader mechanism supporting UEFI secure boot, UKIs and so on.

IIUC for now the only supported boot sequence is `EDK2 -> L4TLauncher -> systemd-boot`, as we have to keep L4TLauncher, because it handles Tegra-specific (so-called "boot chains") A/B redundancy implementation (btw, is there any else I might have missed in L4TLauncher sources that might be useful for Linux booting?).

I was wondering if it's possible ti get rid of `L4TLauncher` completely, and have `systemd-boot` as default boot file `EFI/BOOT/BOOTAA64.efi`. EDK2 `BootChainDxe` driver instead of setting EFI variables `BootChainFwCurrent`/`BootChainOsCurrent` can directly "flip" boot entries (considering we have entries like `default-slot.conf` and `fallback-slot.conf`) in ESP partition. I'm not sure if it's a good idea, but I haven't found any restrictions in UEFI PI spec

**Comment by dwalkes** [comment]
> I was wondering if it's possible to get rid of L4TLauncher completely,

I don't have any direct knowledge in this area but there is also a discussion about [using grub instead of L4TLauncher](https://docs.nvidia.com/jetson/archives/r36.3/DeveloperGuide/SD/Bootloader/UEFI.html#grub-support) at https://github.com/OE4T/meta-tegra/discussions/1480 if you haven't found this already.  Perhaps this implementation could be useful to figure out how to replace l4tlauncher in a systemd-boot scheme.

**Comment by dwalkes** [comment]
This was briefly discussed in the meeting this month at https://github.com/OE4T/meta-tegra/wiki/OE4T-Meeting-Notes-2024%E2%80%9009%E2%80%9012 as well (see the end of the meeting)
